File: core/src/main/cpp/userauth/UserAuthenticationScheme.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <userauth/UserAuthenticationScheme.h>

using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace userauth {

//StaticMslMutex UserAuthenticationScheme::mutex_;
//std::map<std::string, UserAuthenticationScheme> UserAuthenticationScheme::schemes();

const UserAuthenticationScheme UserAuthenticationScheme::EMAIL_PASSWORD("EMAIL_PASSWORD");
const UserAuthenticationScheme UserAuthenticationScheme::USER_ID_TOKEN("USER_ID_TOKEN");
const UserAuthenticationScheme UserAuthenticationScheme::INVALID("INVALID");

// static
IMutex& UserAuthenticationScheme::mutex()
{
    static StaticMslMutex mutex;
    return mutex;
}
    
// static
std::map<std::string, UserAuthenticationScheme>& UserAuthenticationScheme::schemes()
{
    static std::map<std::string, UserAuthenticationScheme> schemes;
    return schemes;
}

// static
UserAuthenticationScheme UserAuthenticationScheme::getScheme(const std::string& name)
{
    LockGuard lockGuard(mutex());
    const MapType::const_iterator it = schemes().find(name);
    return (it == schemes().end()) ? INVALID : it->second;
}

// static
std::vector<UserAuthenticationScheme> UserAuthenticationScheme::values()
{
    LockGuard lockGuard(mutex());
    std::vector<UserAuthenticationScheme> v;
    for(MapType::const_iterator it = schemes().begin(); it != schemes().end(); ++it)
        if (it->second != INVALID) v.push_back(it->second);
    return v;
}

UserAuthenticationScheme::UserAuthenticationScheme(const std::string& name) : name_(name)
{
    LockGuard lockGuard(UserAuthenticationScheme::mutex());
    schemes().insert(std::make_pair(name, *this));
}

bool operator==(const UserAuthenticationScheme& a, const UserAuthenticationScheme& b)
{
    return a.name() == b.name();
}

bool operator!=(const UserAuthenticationScheme& a, const UserAuthenticationScheme& b)
{
    return !(a == b);
}

bool operator<(const UserAuthenticationScheme& a, const UserAuthenticationScheme& b) {
    return a.name() > b.name();
}

}}} // namespace netflix::msl::userauth

File: core/src/main/javascript/userauth/UserAuthenticationScheme.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>User authentication schemes.</p>
 * 
 * <p>The scheme name is used to uniquely identify user authentication
 * schemes.</p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";

	var Class = require('../util/Class.js');
    
    /** Map of names onto schemes. */
    var schemes = {};
    
    var UserAuthenticationScheme = module.exports = Class.create({
        /**
         * Define a user authentication scheme with the specified name.
         * 
         * @param {string} name the user authentication scheme name.
         */
        init: function init(name) {
            // The properties.
            var props = {
                name: { value: name, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
    
            // Add this scheme to the map.
            schemes[name] = this;
        },

        /** @inheritDoc */
        toString: function toString() {
            return this.name;
        },
    });
    
    Class.mixin(UserAuthenticationScheme,
    /** @lends UserAuthenticationScheme */
    ({
        /** Email/password. */
        EMAIL_PASSWORD : new UserAuthenticationScheme("EMAIL_PASSWORD"),
        USER_ID_TOKEN : new UserAuthenticationScheme("USER_ID_TOKEN"),
    }));

    /**
     * @param {string} name the entity authentication scheme name.
     * @return {?UserAuthenticationScheme} the scheme identified by the specified name or {@code null} if
     *         there is none.
     */
    var UserAuthenticationScheme$getScheme = function UserAuthenticationScheme$getScheme(name) {
        return (schemes[name]) ? schemes[name] : null;
    };
    
    // Exports.
    Object.defineProperties(UserAuthenticationScheme, {
        getScheme: { value: UserAuthenticationScheme$getScheme, writable: false, enumerable: false, configurable: false },
    });
    Object.freeze(UserAuthenticationScheme);
})(require, (typeof module !== 'undefined') ? module : mkmodule('UserAuthenticationScheme'));